src/rust_connection/id_allocator.rs

```diff
@@ -42,24 +44,38 @@ impl IDAllocator {
         conn: &impl RequestConnection,
     ) -> Result<u32, ReplyOrIdError> {
         if self.next_id > self.max_id {
-            // TODO: Check if XC-MISC is available.
-
-            // Send an XC-MISC GetXIDRange request.
-            let xidrange = conn.xc_misc_get_xidrange()?.reply()?;
-            let (start, count) = (xidrange.start_id, xidrange.count);
-            // Apparently (0, 1) is how the server signals "I am out of IDs".
-            // The second case avoids an underflow below and should never happen.
-            if (start, count) == (0, 1) || count == 0 {
+            if self.check_xc_misc(conn)? {
+                // Send an XC-MISC GetXIDRange request.
+                let xidrange = conn.xc_misc_get_xidrange()?.reply()?;
+                let (start, count) = (xidrange.start_id, xidrange.count);
+                // Apparently (0, 1) is how the server signals "I am out of IDs".
+                // The second case avoids an underflow below and should never happen.
+                if (start, count) == (0, 1) || count == 0 {
+                    return Err(ReplyOrIdError::IdsExhausted);
+                }
+                self.next_id = start;
+                self.max_id = start + (count - 1) * self.increment;
+            } else {
                 return Err(ReplyOrIdError::IdsExhausted);
             }
-            self.next_id = start;
-            self.max_id = start + (count - 1) * self.increment;
         }
         assert!(self.next_id <= self.max_id);
         let id = self.next_id;
         self.next_id += self.increment;
         Ok(id)
     }
+
+    /// Checks whether XC-MISC is available, caching the result
+    fn check_xc_misc(&mut self, conn: &impl RequestConnection) -> Result<bool, ReplyError> {
+        if let Some(has_xc_misc) = self.has_xc_misc {
+            return Ok(has_xc_misc);
+        }
+
+        let ext_info = conn.extension_information("XC-MISC")?;
+        let has_xc_misc = ext_info.is_some();
+        self.has_xc_misc = Some(true);
```

Comment:
That's definitely not what you had in mind here.
```suggestion
        self.has_xc_misc = Some(has_xc_misc);
```

Reply:
oops!